File: JavaScript/ExampleTests/common/AddressString.js
import * as tokens from "./tokens.js";

export class AddressString {
  static DNC_STRING = "?DNC?";

  constructor(street = "", city = "", state = "", zIP = "") {
    this.street = street;
    this.city = city;
    this.state = state;
    this.zIP = zIP;
  }

  static fromList(values) {
    const v = Array.from(values);
    return new AddressString(
      v[0] ?? "",
      v[1] ?? "",
      v[2] ?? "",
      v[3] ?? ""
    );
  }

  /** Builds from the text form, e.g. Money as "25 USD". */
  static fromText(text) {
    const parts = tokens.require_(text, 4, "Address");
    return new AddressString(
      parts[0],
      parts[1],
      parts[2],
      parts[3]
    );
  }

  toString() {
    return tokens.token(this.street) + " " + tokens.token(this.city) + " " + tokens.token(this.state) + " " + tokens.token(this.zIP);
  }

  equals(other) {
    if (!(other instanceof AddressString)) return false;
    return (this.street === AddressString.DNC_STRING || other.street === AddressString.DNC_STRING || this.street === other.street)
      && (this.city === AddressString.DNC_STRING || other.city === AddressString.DNC_STRING || this.city === other.city)
      && (this.state === AddressString.DNC_STRING || other.state === AddressString.DNC_STRING || this.state === other.state)
      && (this.zIP === AddressString.DNC_STRING || other.zIP === AddressString.DNC_STRING || this.zIP === other.zIP);
  }

  static equalLists(a, b) {
    if (a.length !== b.length) return false;
    return a.every((row, i) => row.equals(b[i]));
  }
}

File: Java/SpectableJavaExamplesTest/src/test/java/spectable/common/AddressString.java
package spectable.common;

import java.util.Objects;
import production.*;
import records.*;
import calculator.*;

public class AddressString {
    private static final String DNCString = "?DNC?";

    public String street;
    public String city;
    public String state;
    public String zIP;

    public AddressString(String street, String city, String state, String zIP) {
        this.street = street;
        this.city = city;
        this.state = state;
        this.zIP = zIP;
    }

    /** Builds from the text form, e.g. Money as "25 USD". */
    public static AddressString fromText(String text) {
        java.util.List<String> parts = Tokens.require(text, 4, "Address");
        return new AddressString(parts.get(0), parts.get(1), parts.get(2), parts.get(3));
    }

    public AddressString(String text) {
        AddressString parsed = fromText(text);
        this.street = parsed.street;
        this.city = parsed.city;
        this.state = parsed.state;
        this.zIP = parsed.zIP;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof AddressString)) return false;
        AddressString that = (AddressString) o;
        return (DNCString.equals(street) || DNCString.equals(that.street) || Objects.equals(street, that.street))
            && (DNCString.equals(city) || DNCString.equals(that.city) || Objects.equals(city, that.city))
            && (DNCString.equals(state) || DNCString.equals(that.state) || Objects.equals(state, that.state))
            && (DNCString.equals(zIP) || DNCString.equals(that.zIP) || Objects.equals(zIP, that.zIP));
    }

    @Override
    public int hashCode() {
        return Objects.hash(street, city, state, zIP);
    }

    @Override
    public String toString() {
        return Tokens.token(street) + " " + Tokens.token(city) + " " + Tokens.token(state) + " " + Tokens.token(zIP);
    }
}
